src/main/scala/org/constellation/snapshotstreaming/SnapshotProcessor.scala

```diff
@@ -0,0 +1,141 @@
+import org.tessellation.security.{Hashed, SecurityProvider}
+
+import com.sksamuel.elastic4s.ElasticDsl.bulk
+import fs2.Stream
+import org.constellation.snapshotstreaming.opensearch.{OpensearchDAO, UpdateRequestBuilder}
+import org.constellation.snapshotstreaming.s3.S3DAO
+import org.http4s.ember.client.EmberClientBuilder
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait SnapshotProcessor[F[_]] {
+  val runtime: Stream[F, Unit]
+}
+
+object SnapshotProcessor {
+
+  def make[F[_]: Async: KryoSerializer: SecurityProvider: Random](configuration: Configuration): Resource[F, SnapshotProcessor[F]] =
+    for {
+      client <- EmberClientBuilder
+        .default[F]
+        .withTimeout(configuration.httpClientTimeout)
+        .withIdleTimeInPool(configuration.httpClientIdleTime)
+        .build
+      opensearchDAO <- OpensearchDAO.make[F](configuration.opensearchUrl)
+      s3DAO <- S3DAO.make[F](configuration)
+      l0GlobalSnapshotClient = L0GlobalSnapshotClient.make[F](client)
+      l0ClusterStorage <- Resource.eval {
+        Ref.of[F, NonEmptyMap[PeerId, L0Peer]](configuration.l0Peers).map(L0ClusterStorage.make(_))
+      }
+      lastSnapshotStorage <- Resource.eval {
+        FileBasedLastGlobalSnapshotStorage.make[F](configuration.lastSnapshotPath)
+      }
+      l0Service = L0Service.make[F](l0GlobalSnapshotClient, l0ClusterStorage, lastSnapshotStorage, configuration.pullLimit.some)
+      requestBuilder = UpdateRequestBuilder.make[F](configuration)
+    } yield make(configuration, lastSnapshotStorage, l0Service, opensearchDAO, s3DAO, requestBuilder)
+
+  def make[F[_]: Async: KryoSerializer: SecurityProvider: Random](
+    configuration: Configuration,
+    lastGlobalSnapshotStorage: LastGlobalSnapshotStorage[F],
+    l0Service: L0Service[F],
+    opensearchDAO: OpensearchDAO[F],
+    s3DAO: S3DAO[F],
+    updateRequestBuilder: UpdateRequestBuilder[F]
+  ): SnapshotProcessor[F] = new SnapshotProcessor[F] {
+
+    val logger = Slf4jLogger.getLogger[F]
+
+    private def prepareAndExecuteBulkUpdate(snapshot: Hashed[GlobalSnapshot]): F[Unit] =
+      Clock[F].realTime
+        .map(d => new Date(d.toMillis))
+        .flatMap(updateRequestBuilder.bulkUpdateRequests(snapshot, _))
+        .flatMap(_.traverse(br => opensearchDAO.sendToOpensearch(bulk(br).refreshImmediately)))
+        .flatMap(_ => logger.info(s"Snapshot ${snapshot.ordinal.value.value} (hash: ${snapshot.hash.show.take(8)}) sent to opensearch."))
+        .void
+
+    private def process(snapshot: Hashed[GlobalSnapshot]): F[Unit] =
+      lastGlobalSnapshotStorage.get.flatMap {
+        case Some(last) if Validator.isNextSnapshot(last, snapshot) =>
+          s3DAO.uploadSnapshot(snapshot) >>
+            prepareAndExecuteBulkUpdate(snapshot) >>
+            lastGlobalSnapshotStorage.set(snapshot)
+
+        case Some(last) =>
+          logger.warn(s"Pulled snapshot doesn't form a correct chain, ignoring! Last: ${GlobalSnapshotReference
```

Comment:
will it try fetching from another node?

Reply:
Yes it works similarly to how it works in dag L1. The pulling will execute again in couple of seconds (configured with `pullInterval`) and it will try to fetch the next snapshots after the last accepted one. Wrong snapshots are not accepted, so the retry will happen from a randomly picked node.